### PR TITLE
Fixed divrem bugs, added pseudo divrem

### PR DESCRIPTION
**Pull request by Ovascos**
- added assertion to warn about incorrect dereference in function `coefficient_div` when dividing a constant by a non-constant polynomial
- fixed incorrect D output parameter in `coefficient_divrem` when calling with two constant polynomials
- use dense pseudo remainder in `polynomial_reduce`. Previously it expected lc divisibility rendering it just another divrem function
- added `lp_polynomial_pdivrem` and `lp_polynomial_spdivrem` for (sparse) pseudo division with remainder and added python bindings for it
- removed unimplemented function prototype `lp_polynomial_context` from polynomial.h
- update trace.h to avoid some potential preprocessor replacement pitfalls (used `do {} while(0)` for multiple lines, use `((void)0)` instead of empty command in release builds)
- Fixed algebraic number test, added return value for check to indicate failed tests
- fixed some typos

### CODE REVIEW

File: src/polynomial/polynomial.c [hunk 1]
```diff
@@ -669,7 +719,7 @@ void lp_polynomial_divrem(lp_polynomial_t* D, lp_polynomial_t* R, const lp_polyn
   lp_polynomial_set_context(D, A1->ctx);
   lp_polynomial_set_context(R, A1->ctx);
 
-  coefficient_divrem(D->ctx, &R->data, &R->data, &A1->data, &A2->data);
+  coefficient_divrem(D->ctx, &D->data, &R->data, &A1->data, &A2->data);
```
Comment (ahmed-irfan) [inline src/polynomial/polynomial.c:722]: Nice catch

File: include/polynomial.h [hunk 2]
```diff
@@ -220,7 +220,7 @@ void lp_polynomial_sub_mul(lp_polynomial_t* S, const lp_polynomial_t* A1, const
  *
  * and
  *
- *   P in Z[y]
+ *   P in Z[x]
```
Comment (ahmed-irfan) [inline include/polynomial.h:220]: This conflicts with the comment of the coefficient_reduce method. Are you sure about this?

Comment (Ovascos) [reply include/polynomial.h:220]: I forgot to change this at coefficient_reduce as well.

File: test/python/tests/algebraic_number.py [hunk 3]
```diff
@@ -159,4 +159,4 @@ def check_comparison(a1, a2, cmp, result, expected):
     p = functools.reduce(lambda x, y: x*y, sample, one)
     random.shuffle(sample)
     p = functools.reduce(lambda x, y: x/y, sample, p)
-    polypy_test.check(p == 1)
+    polypy_test.check(p == one)
```
Comment (ahmed-irfan) [inline test/python/tests/algebraic_number.py:162]: nice catch

**Review by ahmed-irfan — APPROVED** [review]
LGTM!

Thanks

**Pull request merged**